File: chemistry/chemistry_mpi2.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6370
! 9 January 2009
!=================================================================


program ChemicalEquilibrium_MPI
  !-----------------------------------------------------------------
  ! Description: 
  !    Computes the equilibrium chemical densities at a number of 
  !    spatial locations, given a (random) background temperature 
  !    field.  The chemical rate equations and solution strategy
  !    are in the subroutine chem_solver, which is called at every
  !    spatial location.
  !-----------------------------------------------------------------
  !======= Inclusions ===========
  use mpi
!  include "mpif.h"  ! include this if mpi module unavailable
  implicit none

  !======= Interfaces ===========
  interface
     subroutine chem_solver(T,u,v,w,lam,eps,maxit,its,res)
       integer, intent(in) :: maxit
       integer, intent(out) :: its
       double precision, intent(in) :: T, lam, eps
       double precision, intent(out) :: res
       double precision, intent(inout) :: u, v, w
     end subroutine chem_solver
  end interface

  !======= Declarations =========
  integer :: n, i, maxit, its, numprocs, myid, ierr, mycount
  integer :: numsent, sender, ansentry, status(MPI_STATUS_SIZE)
  double precision, allocatable :: T(:), u(:), v(:), w(:)
  double precision :: Tval, uval, vval, wval
  double precision :: buffer(3)
  double precision :: lam, eps, res, stime, ftime
  logical :: more_work
  
  !======= Internals ============
  
  ! set solver input parameters
  maxit = 1000000
  lam = 1.d-2
  eps = 1.d-10

  ! initialize internal work counter
  mycount = 0
  
  ! intialize MPI
  call mpi_init(ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_init =',ierr
     stop
  endif
  call mpi_comm_size(MPI_COMM_WORLD, numprocs, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_comm_size =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif
  call mpi_comm_rank(MPI_COMM_WORLD, myid, ierr)
  if (ierr /= 0) then
     write(*,*) ' error in mpi_comm_rank =',ierr
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif

  ! check whether we have enough processors
  if (numprocs < 2) then
     write(*,*) ' master/slave example, requires > 1 proc!  Exiting'
     call mpi_abort(MPI_COMM_WORLD, -1, ierr)
     stop
  endif

  ! master role
  if (myid == 0) then

     ! input the total number of intervals
     write(*,*) 'Enter the total number of intervals (0 quits):'
     read(*,*) n
     write(*,'(A,i5,A)') ' starting MPI with ', numprocs,' processes'

     ! stop for illegal n
     if (n < 1) then
        call mpi_abort(MPI_COMM_WORLD, -1, ierr)
        stop
     endif

     ! allocate temperature field and solution arrays
     allocate(T(n),u(n),v(n),w(n))

     ! set random temperature field, initial guesses at chemical densities
     call random_number(T)

     ! start timer
     stime = MPI_Wtime()

     ! do master/slave computation (assumes n >= numprocs)
     numsent = 0

     ! first send a part to each slave
     do i = 1,min(n,numprocs-1)
        ! fill send buffer 
        buffer(1) = T(i)
        ! send with tag as entry
        call mpi_send(buffer, 1, MPI_DOUBLE_PRECISION, i, i, MPI_COMM_WORLD, ierr)
        if (ierr /= 0) then
           write(*,*) ' error in mpi_send =',ierr
           call mpi_abort(MPI_COMM_WORLD, -1, ierr)
           stop
        endif
        numsent = numsent + 1
     enddo

     ! receive work from each proc, and assign additional work if available
     do i = 1,n

        ! receive answers from any processor
        call mpi_recv(buffer, 3, MPI_DOUBLE_PRECISION, MPI_ANY_SOURCE, &
                      MPI_ANY_TAG, MPI_COMM_WORLD, status, ierr)
        if (ierr /= 0) then
           write(*,*) ' error in mpi_recv =',ierr
           call mpi_abort(MPI_COMM_WORLD, -1, ierr)
           stop
        endif
        ! decode the sender and solution entry from status
        sender = status(MPI_SOURCE)
        ansentry = status(MPI_TAG)
        ! store results
        u(ansentry) = buffer(1)
        v(ansentry) = buffer(2)
        w(ansentry) = buffer(3)
        if (numsent < n) then  ! send another row
           buffer(1) = T(numsent+1)
           call mpi_send(buffer, 1, MPI_DOUBLE_PRECISION, sender, &
                         numsent+1, MPI_COMM_WORLD, ierr)
           if (ierr /= 0) then
              write(*,*) ' error in mpi_send =',ierr
              call mpi_abort(MPI_COMM_WORLD, -1, ierr)
              stop
           endif
           numsent = numsent + 1
        else   ! tell senders that work is complete
           call mpi_send(MPI_BOTTOM, 0, MPI_DOUBLE_PRECISION, sender, 0, &
                         MPI_COMM_WORLD, ierr)
           if (ierr /= 0) then
              write(*,*) ' error in mpi_send =',ierr
              call mpi_abort(MPI_COMM_WORLD, -1, ierr)
              stop
           endif
        endif
           
     enddo

     ! stop timer
     ftime = MPI_Wtime()

     ! free temperature/solution arrays
     deallocate(T,u,v,w)

  else  ! worker role

     ! loop over assignments from master
     more_work = .true.
     do while(more_work) 

        ! receive message from master
        call mpi_recv(buffer, 1, MPI_DOUBLE_PRECISION, 0, MPI_ANY_TAG, &
                      MPI_COMM_WORLD, status, ierr)
        if (ierr /= 0) then
           write(*,*) ' error in mpi_recv =',ierr
           call mpi_abort(MPI_COMM_WORLD, -1, ierr)
           stop
        endif

        ! check if work is complete
        if (status(MPI_TAG) == 0) then
           more_work = .false.

        ! otherwise, do computation
        else

           ! set row index
           i = status(MPI_TAG)

           ! set input and initial guess
           Tval = buffer(1)
           uval = 0.35d0
           vval = 0.1d0
           wval = 0.5d0
           
           ! call solver
           call chem_solver(Tval,uval,vval,wval,lam,eps,maxit,its,res)
           if (res < eps) then
!!$              write(*,'(2(A,i8))') '    i =',i,',  its =',its
           else
              write(*,'(2(A,i8),5(A,es9.2))') '    error: i =',i,',  its =',its, &
                   ',  res =',res,',  T =',Tval,',  u =',uval,',  v =',vval,',  w =',wval
              call mpi_abort(MPI_COMM_WORLD, -1, ierr)
           endif
           
           ! send result to master
           buffer(1) = uval
           buffer(2) = vval
           buffer(3) = wval
           call mpi_send(buffer, 3, MPI_DOUBLE_PRECISION, 0, i, MPI_COMM_WORLD, ierr)
           if (ierr /= 0) then
              write(*,*) ' error in mpi_send =',ierr
              call mpi_abort(MPI_COMM_WORLD, -1, ierr)
              stop
           endif
           
           ! update work counter
           mycount = mycount + 1

        endif

     end do

  endif

  ! output solution time
  write(*,'(A,i3,A,i5,A)') ' proc ',myid,' computed ',mycount,' iterations'
  if (myid == 0) then
     write(*,*) '     runtime =',ftime-stime
  endif


  ! finalize MPI
  call mpi_finalize(ierr)  

end program ChemicalEquilibrium_MPI
!=================================================================

File: chemistry/chemistry_mpi2.cpp
/* Daniel R. Reynolds
   SMU Mathematics
   Math 6370
   4 January 2013 */

// Inclusions
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "mpi.h"

// Prototypes
void chem_solver(double, double*, double*, double*, 
		 double, double, int, int*, double*);


/* Example routine to compute the equilibrium chemical densities at 
   a number of spatial locations, given a (random) background temperature
   field.  The chemical rate equations and solution strategy are in the 
   subroutine chem_solver, which is called at every spatial location. */
int main(int argc, char* argv[]) {

  // declarations
  int maxit, n, i, its, iend, numsent, numreceived, mycount;
  double lam, eps, *buffer, res, runtime;
  double stime, ftime;
  int ierr, numprocs, myid, sender, ansentry;
  MPI_Status status;
  bool more_work;

  // intialize MPI
  ierr = MPI_Init(&argc, &argv);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Init = %i\n",ierr);
    return 1;
  }
  ierr = MPI_Comm_size(MPI_COMM_WORLD, &numprocs);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Comm_size = %i\n",ierr);
    return 1;
  }
  ierr = MPI_Comm_rank(MPI_COMM_WORLD, &myid);
  if (ierr != MPI_SUCCESS) {
    printf("Error in MPI_Comm_rank = %i\n",ierr);
    return 1;
  }

  // check whether we have enough processes
  if (numprocs < 2) {
    printf("manager/worker example, requires > 1 proc! Exiting\n");
    ierr = MPI_Abort(MPI_COMM_WORLD, 1);
    return 1;
  }

  // set solver parameters, temporary variables
  maxit = 1000000;
  lam = 1.e-2;
  eps = 1.e-10;
  buffer = new double[3];


  /////////////////
  // manager role
  if (myid == 0) {

    printf("Starting MPI with %i processes\n\n",numprocs);

    // input the total number of intervals
    printf("Enter the total number of intervals (0 quits):\n");
    i = scanf("%i", &n);
    if (i < 1) {
      ierr = MPI_Finalize();
      return 1;
    }

    // stop for illegal n
    if (n < 1) {
      ierr = MPI_Finalize();
      return 1;
    }

    // allocate temperature field and solution arrays
    double *T = new double[n];
    double *u = new double[n];
    double *v = new double[n];
    double *w = new double[n];

    // set random temperature field
    for (i=0; i<n; i++)  
      T[i] = random() / (pow(2.0,31.0) - 1.0);

    // start timer
    stime = MPI_Wtime();

    // do manager/worker computation (assumes n >= numprocs)
    numsent = 0;

    // first send a part to each worker
    iend = (n < numprocs-1) ? n : numprocs-1;
    for (i=0; i<iend; i++) {
      // fill send buffer
      buffer[0] = T[i];
      // send with tag as entry in temperature array
      ierr = MPI_Send(buffer, 1, MPI_DOUBLE, i+1, numsent, MPI_COMM_WORLD);
      if (ierr != 0) {
	printf("Error in MPI_Send = %i\n",ierr);
	ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	return 1;
      }
      numsent++;
    } // for i

    // receive work from each proc, and assign additional work if available
    for (i=0; i<n; i++) {

      // receive answers from any process
      ierr = MPI_Recv(buffer, 3, MPI_DOUBLE, MPI_ANY_SOURCE, 
		      MPI_ANY_TAG, MPI_COMM_WORLD, &status);
      if (ierr != 0) {
	printf("Error in MPI_Recv = %i\n",ierr);
	ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	return 1;
      }
      // decode the sender and solution entry from status
      sender   = status.MPI_SOURCE;
      ansentry = status.MPI_TAG;
      // store results
      u[ansentry] = buffer[0];
      v[ansentry] = buffer[1];
      w[ansentry] = buffer[2];
      if (numsent < n) {  // send another row
	buffer[0] = T[numsent];
	ierr = MPI_Send(buffer, 1, MPI_DOUBLE, sender, numsent, MPI_COMM_WORLD);
	if (ierr != 0) {
	  printf("Error in MPI_Send = %i\n",ierr);
	  ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	  return 1;
	}
	numsent++;
      } else {  // tell senders that work is complete, by sending message of zero size
	ierr = MPI_Send(buffer, 0, MPI_DOUBLE, sender, 0, MPI_COMM_WORLD);
	if (ierr != 0) {
	  printf("Error in MPI_Send = %i\n",ierr);
	  ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	  return 1;
	}
      } // if numsent
           
    } // for i

    // stop timer
    ftime = MPI_Wtime();
    runtime = ftime - stime;

    // output runtime
    printf("     runtime = %.16e\n",runtime);

    // delete temperature and solution arrays
    delete[] T;
    delete[] u;
    delete[] v;
    delete[] w;


  /////////////////
  // worker role
  } else {
    
    // allocate temperature and solution values to reuse
    double T, u, v, w;

    // initialize internal work counter
    mycount = 0;

    // loop over assignments from manager
    more_work = true;
    while (more_work) {

      // receive message from manager
      ierr = MPI_Recv(buffer, 1, MPI_DOUBLE, 0, MPI_ANY_TAG, 
		      MPI_COMM_WORLD, &status);
      if (ierr != 0) {
	printf("Error in MPI_Recv = %i\n",ierr);
	ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	return 1;
      }

      // check if work is complete
      ierr = MPI_Get_count(&status, MPI_DOUBLE, &numreceived);
      if (numreceived == 0) {
	more_work = false;

      // otherwise, do computation
      } else {

	// set row index
	i = status.MPI_TAG;

	// set input and initial guess
	T = buffer[0];
	u = 0.35;
	v = 0.1;
	w = 0.5;

	// call solver
	chem_solver(T, &u, &v, &w, lam, eps, maxit, &its, &res);
	if (res < eps) {
 	  // printf("    i = %i,  its = %i\n", i, its);
	}
	else {
	  printf("    error: i=%i, its=%i, res=%.2e, u=%.2e, v=%.2e, w=%.2e\n", 
		 i, its, res, u, v, w);
	  ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	  return 1;
	}

	// send result to manager
	buffer[0] = u;
	buffer[1] = v;
	buffer[2] = w;
	ierr = MPI_Send(buffer, 3, MPI_DOUBLE, 0, i, MPI_COMM_WORLD);
	if (ierr != 0) {
	  printf("Error in MPI_Send = %i\n",ierr);
	  ierr = MPI_Abort(MPI_COMM_WORLD, 1);
	  return 1;
	}

	// update work counter
	mycount++;

      } // if status.MPI_TAG

    } // end while loop

    // output work performed
    printf(" proc %i computed %i iterations\n", myid, mycount);

  } // end worker section

  // finalize MPI
  ierr = MPI_Finalize();

} // end main
